src/librustc/ty/sty.rs

```diff
@@ -2387,8 +2387,11 @@ impl<'tcx> Const<'tcx> {
             let (param_env, substs) = param_env_and_substs.into_parts();
 
             // try to resolve e.g. associated constants to their definition on an impl, and then
-            // evaluate the const.
-            tcx.const_eval_resolve(param_env, did, substs, promoted, None).ok()
+            // evaluate the const. We don't use the type from the evaluated const as that has had it
+            // lifetimes erased.
+            tcx.const_eval_resolve(param_env, did, substs, promoted, None)
```

Comment:
Btw, this code is in a closure called `try_const_eval` because it's used twice below, and one of those is a workaround for having inference variables in the `substs`.

So all uses of `ty::Const::eval` which are from an inference context, should probably be replaced with a version that is simpler (because none of the `has_local_value` checks would be necessary) and which goes through `InferCtxt::const_eval_resolve`.

Reply:
I'm going to do this in a follow up PR, as there is quite a few place that use `Const::eval` methods.

Reply:
Makes sense, can you leave a FIXME comment somewhere in this code?